File: backend/workline/knowledge/cache/invalidation.py
"""Targeted invalidation manager tracking project and source dependencies."""

import threading
from typing import Dict, List, Set


class InvalidationManager:
    """Maintains reverse indices mapping source IDs and project IDs to cache keys."""

    def __init__(self):
        self._lock = threading.RLock()
        self._source_to_keys: Dict[str, Set[str]] = {}
        self._project_to_keys: Dict[str, Set[str]] = {}

    def track(self, key: str, project_id: str, source_id: str = None) -> None:
        with self._lock:
            if project_id:
                if project_id not in self._project_to_keys:
                    self._project_to_keys[project_id] = set()
                self._project_to_keys[project_id].add(key)

            if source_id:
                if source_id not in self._source_to_keys:
                    self._source_to_keys[source_id] = set()
                self._source_to_keys[source_id].add(key)

    def get_keys_by_source(self, source_id: str) -> List[str]:
        with self._lock:
            return list(self._source_to_keys.get(source_id, set()))

    def get_keys_by_project(self, project_id: str) -> List[str]:
        with self._lock:
            return list(self._project_to_keys.get(project_id, set()))

    def remove_key(self, key: str) -> None:
        with self._lock:
            for s in self._source_to_keys.values():
                s.discard(key)
            for s in self._project_to_keys.values():
                s.discard(key)

    def clear(self) -> None:
        with self._lock:
            self._source_to_keys.clear()
            self._project_to_keys.clear()

File: packages/knowledge/cache/invalidation.ts
/**
 * Invalidation engine tracking dependencies and project invalidation rules.
 */

import { CacheEntry, CacheObjectType } from "./metadata";

export class InvalidationEngine {
  private sourceToKeys: Map<string, Set<string>> = new Map();
  private projectToKeys: Map<string, Set<string>> = new Map();

  public track(entry: CacheEntry): void {
    const key = entry.metadata.cacheKey;
    const proj = entry.metadata.projectId;
    const src = entry.metadata.sourceId;

    if (proj) {
      if (!this.projectToKeys.has(proj)) {
        this.projectToKeys.set(proj, new Set());
      }
      this.projectToKeys.get(proj)!.add(key);
    }

    if (src) {
      if (!this.sourceToKeys.has(src)) {
        this.sourceToKeys.set(src, new Set());
      }
      this.sourceToKeys.get(src)!.add(key);
    }
  }

  public getKeysBySource(sourceId: string): string[] {
    const set = this.sourceToKeys.get(sourceId);
    return set ? Array.from(set) : [];
  }

  public getKeysByProject(projectId: string): string[] {
    const set = this.projectToKeys.get(projectId);
    return set ? Array.from(set) : [];
  }

  public removeKey(key: string): void {
    for (const set of this.sourceToKeys.values()) {
      set.delete(key);
    }
    for (const set of this.projectToKeys.values()) {
      set.delete(key);
    }
  }

  public clear(): void {
    this.sourceToKeys.clear();
    this.projectToKeys.clear();
  }
}
